lib/queue/events.ts

```diff
@@ -260,34 +131,32 @@ export const wait = async (
 		stream.once('closed', () => {
 			stream.removeAllListeners();
 			logger.info(logContext, 'Closing wait stream', {
-				slug,
+				id: actionRequestId,
 			});
 
 			resolve(result);
 		});
 
-		// Don't perform a "get by slug" if we already have a match.
+		// Don't attempt a direct query if we already have a match.
 		if (result) {
 			return;
 		}
 
 		kernel
-			.getContractBySlug(
-				logContext,
-				session,
-				`${slug}@${EXECUTION_EVENT_VERSION}`,
-			)
-			.then((contract) => {
-				if (!contract) {
+			.query(logContext, session, schema, {
+				limit: 1,
+			})
+			.then((contracts) => {
+				if (!contracts[0]) {
 					return;
```

Comment:
Should we be calling `stream.close();` here? That's the sort of thing that seems like it should always be called regardless of how the function is exited. I'm assuming the answer is actually no because if the contract is not found you want the stream to continue waiting for it to potentially be created but that's not immediately obvious (and if that is the case it's probably worth adding a comment to make it clear to people who might later try to come along and "fix" the assumed omission)

Reply:
You're correct, we want to continue waiting if a matching contract isn't immediately found. I'll go ahead and add a comment, but this made me think, should we only open a stream if an immediate query doesn't find a match? Seems like it could save us from opening unnecessary streams in some cases. I'll test out my hypothesis and report back.

Reply:
Ok, ran all tests after opening a stream only after a query returns empty results and everything passed. That's makes more logical sense to me, what do you think?

Reply:
Ah, I see what you mean and I think you're right, I'll revert my changes.